### DISCUSSION TITLE
How to log output something like a structured dict object instead of a serialized string

### DISCUSSION THREAD

**Discussion by mirumirumi** [aws-powertools/powertools-lambda-python#2086] [Q&A] [answered]
I am currently working on Rust on Lambda.
I am used to developing in Python & PowerTools and would like to produce similar structured logs in Rust.

However, when I started learning about loggers, I learned that normal logging is usually done with values in parallel directly under the `fields` key, and it is not common to log deeply nested objects as they are.

I have been logging dictionaries whose contents I want to look at as shown below, which is actually convenient (in this case, the logs appear in the `message` key, not in the `fields` key, but this is not important to me).

```python
something_deep_dict = {
  ...
  {
     ...
  }
}

logger.info(something_deep_dict)
```

How does PowerTools output structured logs?
Do you have any special implementations?

Thank you!

**Comment by heitorlessa** [comment, marked as answer]
hey @mirumirumi, for Rust, I'd defer to @nmoutschen. For Python, we're not doing any special on that regard other than serializing to `str` [any non-serializable value](https://awslabs.github.io/aws-lambda-powertools-python/2.11.0/core/logger/#custom-function-for-unserializable-values). Also, note that [you can have additional fields at the root](https://awslabs.github.io/aws-lambda-powertools-python/2.11.0/core/logger/#ephemeral-metadata) (in parallel to `message` key), when you pass additional key-value pairs **after** the logging message.

Here's a quick example I ran locally:

```python
from aws_lambda_powertools import Logger
from typing import Any

logger = Logger(service="discussion-2086")

nested_dict: dict[str, Any] = {
    "products": [
        {"id": "8dbc54d1-6e5c-4f9b-8f4c-2a5fd3e2b2a7", "description": "dummy"},
    ],
    "customer_id": "95126722-4665-4a44-8150-8e12233c9bcd",
}

logger.info("Incoming products", some_key="some_value", **nested_dict)
```

**JSON output**

```json
{
    "level": "INFO",
    "location": "<module>:13",
    "message": "Incoming products",
    "timestamp": "2023-04-05 17:14:46,838+0200",
    "service": "discussion-2086",
    "some_key": "some_value",
    "products": [
        {
            "id": "8dbc54d1-6e5c-4f9b-8f4c-2a5fd3e2b2a7",
            "description": "dummy"
        }
    ],
    "customer_id": "95126722-4665-4a44-8150-8e12233c9bcd"
}
```
